File: acceptance/arpan/13.roman_to_integer.cpp
class Solution {
public:
    int romanToInt(string s) {
        int sum=0;
        for(int i=0;i<s.size();i++){
            if(s[i]=='I'&& s[i+1]=='V'){
                sum+=4;
                i++;
            }else if(s[i]=='I'&& s[i+1]=='X'){
                sum+=9;
                i++;
            }else if(s[i]=='X'&& s[i+1]=='L'){
                sum+=40;
                i++;
            }else if(s[i]=='X'&& s[i+1]=='C'){
                sum+=90;
                i++;
            }else if(s[i]=='C'&& s[i+1]=='D'){
                sum+=400;
                i++;
            }else if(s[i]=='C'&& s[i+1]=='M'){
                sum+=900;
                i++;
            }else if(s[i]=='I'){
                sum+=1;
            }else if(s[i]=='V'){
                sum+=5;
            }else if(s[i]=='X'){
                sum+=10;
            }else if(s[i]=='L'){
                sum+=50;
            }else if(s[i]=='C'){
                sum+=100;
            }else if(s[i]=='D'){
                sum+=500;
            }else if(s[i]=='M'){
                sum+=1000;
            }
        }
        return sum;
    }
};

File: acceptance/rls129/13_roman_to_integer.rs
use std::collections::HashMap;

impl Solution {
    pub fn roman_to_int(num: String) -> i32 {
        let mut max = 0;

        // TODO:
        // this can be made much much faster by using a lookup table
        let mut map = HashMap::new();
        map.insert('I', 1);
        map.insert('V', 5);
        map.insert('X', 10);
        map.insert('L', 50);
        map.insert('C', 100);
        map.insert('D', 500);
        map.insert('M', 1000);

        num.chars().rev().map(|xx| {
            let x = map.get(&xx).unwrap().clone();
            if x >= max {
                max = x;
                x
            } else {
                x*-1
            }
        }).sum()
    }
}
